### PR TITLE
Add ilike and nilike operators

### LINKED ISSUE #633: Pass through ilike and nilike parameters so they can be handled by connectors
Every application needs to be able to search without case sensitivity. Other than regular expressions (not always supported) there isn't a good way to do this. See PR for fixing this in Postgres connector (https://github.com/strongloop/loopback-connector-postgresql/pull/93). This patch needs to be applied to the juggler as well.

### PR DESCRIPTION
**Pull request by duffn**
Closes #633

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by duffn** [comment]
Originally talked about here: https://github.com/strongloop/loopback-datasource-juggler/issues/633#issuecomment-247487564

The original issue was closed because `regexp` was added and it was thought that we didn't need `ilike` and `nilike` because `regexp` could handle all situations. However, we should have `ilike` and `nilike` operators as some databases may not support case-insensitive `regexp`.

**Comment by superkhau** [comment]
Can you add some tests to verify your changes and prevent regressions in the future?

**Comment by duffn** [comment]
I added tests for `ilike` and `nilike` which also required that I added those two operators to the memory connector. 

I also added `like` and `nlike` tests that were missing.

File: lib/dao.js [hunk 1]
```diff
@@ -1661,7 +1665,8 @@ DataAccessObject._coerce = function(where) {
           operator = 'regexp';
         } else if (operator === 'regexp' && val instanceof RegExp) {
           // Do not coerce regex literals/objects
-        } else if (!((operator === 'like' || operator === 'nlike') && val instanceof RegExp)) {
+        } else if (!((operator === 'like' || operator === 'nlike' ||
+          operator === 'ilike' || operator === 'nilike') && val instanceof RegExp)) {
```
Comment (superkhau) [inline lib/dao.js]: Please fix indentation. See our style guide at https://github.com/strongloop/loopback-contributor-docs/blob/master/style-guide.md#indentation-of-multi-line-expressions-in-if

File: test/basic-querying.test.js [hunk 2]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
       });
     });
 
+    it('should support "like" that is satisfied', function(done) {
+      User.find({where: {name: {'like': 'John'}}}, function(err, users) {
```
Comment (superkhau) [inline test/basic-querying.test.js]: remove single quotes around `like`

File: test/basic-querying.test.js [hunk 3]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
...
+        should.not.exist(err);
```
Comment (superkhau) [inline test/basic-querying.test.js]: `if (err) return done(err)` for better stack traces.

File: test/basic-querying.test.js [hunk 4]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
...
+        users.should.have.property('length', 1);
```
Comment (superkhau) [inline test/basic-querying.test.js]: users.length.should.equal(1);

File: test/basic-querying.test.js [hunk 5]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
...
+        users[0].name.should.equal('John Lennon');
+        done();
+      });
+    });
+
+    it('should support "like" that is not satisfied', function(done) {
+      User.find({where: {name: {'like': 'Bob'}}}, function(err, users) {
```
Comment (superkhau) [inline test/basic-querying.test.js]: Remove single quotes around `like`

File: test/basic-querying.test.js [hunk 6]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
...
+        should.not.exist(err);
```
Comment (superkhau) [inline test/basic-querying.test.js]: if (err) return done(err);

File: test/basic-querying.test.js [hunk 7]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
...
+        users.should.have.property('length', 0);
```
Comment (superkhau) [inline test/basic-querying.test.js]: users.length.should.equal(1);

File: test/basic-querying.test.js [hunk 8]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
...
+        done();
+      });
+    });
+
+    it('should support "nlike" that is satisfied', function(done) {
+      User.find({where: {name: {'nlike': 'John'}}}, function(err, users) {
+        should.not.exist(err);
```
Comment (superkhau) [inline test/basic-querying.test.js]: ditto

File: test/basic-querying.test.js [hunk 9]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
...
+        users.should.have.property('length', 5);
```
Comment (superkhau) [inline test/basic-querying.test.js]: ditto

File: test/basic-querying.test.js [hunk 10]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
...
+    it('should support "like" that is not satisfied', function(done) {
+      User.find({where: {name: {'like': 'Bob'}}}, function(err, users) {
+        should.not.exist(err);
+        users.should.have.property('length', 0);
+        done();
+      });
+    });
+
+    it('should support "nlike" that is satisfied', function(done) {
+      User.find({where: {name: {'nlike': 'John'}}}, function(err, users) {
```
Comment (superkhau) [inline test/basic-querying.test.js]: ditto

File: test/basic-querying.test.js [hunk 11]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
...
+        should.not.exist(err);
+        users.should.have.property('length', 5);
+        users[0].name.should.equal('Paul McCartney');
+        done();
+      });
+    });
+
+    it('should support "ilike" that is satisfied', function(done) {
+      User.find({where: {name: {'ilike': 'john'}}}, function(err, users) {
```
Comment (superkhau) [inline test/basic-querying.test.js]: ditto

File: test/basic-querying.test.js [hunk 12]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
...
+        should.not.exist(err);
```
Comment (superkhau) [inline test/basic-querying.test.js]: ditto

File: test/basic-querying.test.js [hunk 13]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
...
+        users.should.have.property('length', 1);
```
Comment (superkhau) [inline test/basic-querying.test.js]: ditto

File: test/basic-querying.test.js [hunk 14]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
...
+        users[0].name.should.equal('John Lennon');
+        done();
+      });
+    });
+
+    it('should support "ilike" that is not satisfied', function(done) {
+      User.find({where: {name: {'ilike': 'bob'}}}, function(err, users) {
```
Comment (superkhau) [inline test/basic-querying.test.js]: ditto

File: test/basic-querying.test.js [hunk 15]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
...
+        should.not.exist(err);
```
Comment (superkhau) [inline test/basic-querying.test.js]: ditto

File: test/basic-querying.test.js [hunk 16]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
...
+        users.should.have.property('length', 0);
```
Comment (superkhau) [inline test/basic-querying.test.js]: ditto

File: test/basic-querying.test.js [hunk 17]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
...
+        done();
+      });
+    });
+
+    it('should support "nilike" that is satisfied', function(done) {
+      User.find({where: {name: {'nilike': 'john'}}}, function(err, users) {
```
Comment (superkhau) [inline test/basic-querying.test.js]: ditto

File: test/basic-querying.test.js [hunk 18]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
...
+        should.not.exist(err);
```
Comment (superkhau) [inline test/basic-querying.test.js]: ditto

File: test/basic-querying.test.js [hunk 19]
```diff
@@ -469,6 +469,58 @@ describe('basic-querying', function() {
...
+        users.should.have.property('length', 5);
```
Comment (superkhau) [inline test/basic-querying.test.js]: ditto

File: lib/dao.js (hunk 1 above)
Comment (duffn) [inline lib/dao.js]: I've corrected per the style guide, but the linter does not give any errors or warnings on any of the files.

File: test/basic-querying.test.js (hunk 2 above)
Comment (duffn) [inline test/basic-querying.test.js]: I've removed all single quotes.

File: test/basic-querying.test.js (hunk 3 above)
Comment (duffn) [inline test/basic-querying.test.js]: I've updated all per the suggestion. This does not match the rest of the tests in this file - I can update those in another PR.

File: test/basic-querying.test.js (hunk 4 above)
Comment (duffn) [inline test/basic-querying.test.js]: I've updated all per the suggestion. This does not match the rest of the tests in this file - I can update those in another PR.

File: lib/dao.js (hunk 1 above)
Comment (superkhau) [inline lib/dao.js]: The linter seems to fail to recognize indentation sometimes. See these lines:

>When not feasible, then indent the second and next lines by two levels.
>One level of indentation makes it difficult to tell the difference between the condition and the branch body.

Read the good/bad examples: https://github.com/strongloop/loopback-contributor-docs/blob/master/style-guide.md

File: test/basic-querying.test.js (hunk 3 above)
Comment (superkhau) [inline test/basic-querying.test.js]: Thanks, this is fine. All those other parts of the file need to change to this style anyways (if you would like to submit a PR for that, I can help approve that one too ;)).

Comment (superkhau) [inline test/basic-querying.test.js]: If you decide to go and refactor that file, you can rename the set of tests in this PR too -- the wording is pretty bad -- `it('supports the "like" operator', function(...)` or something along those lines would be better.

**Comment by superkhau** [comment]
test please

**Comment by superkhau** [comment]
@duffn The unfortunate part about this PR is that adding in those operations causes all the downstream connectors to fail (because they have no implementation). Do you have time to submit PRs for those connectors too? This feature should really go in together with the connector updates.

**Comment by duffn** [comment]
Oh my. That's a lot, but I'll consider it. If I do this, is there anything special I should do to coordinate?

**Comment by superkhau** [comment]
>Oh my. That's a lot, but I'll consider it. If I do this, is there anything special I should do to coordinate? 

The ugly thing is that once you create those PRs, we'll need to merge in this PR first before the other PRs pass. I'm reluctant to merge this in until we have PRs for those other connectors too. @bajtos @raymondfeng We really need to do something about this, what should we do in this case? This PR is LGTM, but downstream connectors will fail if I merge.

**Comment by superkhau** [comment]
Related issue for postgres connector: https://github.com/strongloop/loopback-connector-postgresql/pull/93#issuecomment-248625934

**Comment by bajtos** [comment]
> The ugly thing is that once you create those PRs, we'll need to merge in this PR first before the other PRs pass. I'm reluctant to merge this in until we have PRs for those other connectors too. @bajtos @raymondfeng We really need to do something about this, what should we do in this case? This PR is LGTM, but downstream connectors will fail if I merge.

IMO, we need a way for connectors to signal what features they support. The shared test suite in juggler should skip tests for features that are not supported.

**Comment by duffn** [comment]
So regarding the regex discussion here https://github.com/strongloop/loopback-connector-postgresql/pull/93#issuecomment-248779374, that's fine for PostgreSQL and likely many databases, but I'm writing a connector for Redshift and it does not support case-insensitive regex, however, it does support `ilike`, hence this PR.

**Comment by Amir-61** [comment]
@duffn @superkhau 

>IMO, we need a way for connectors to signal what features they support. The shared test suite in juggler should skip tests for features that are not supported.

I believe you can do similar approach to what @bajtos and I did in [PR#866](https://github.com/strongloop/loopback-datasource-juggler/pull/866).


CC: @bajtos

**Comment by superkhau** [comment]
@Amir-61 I agree with the land and add unsupported test case approach for now. We'll need to revisit how connectors broadcast what they support as stated by @bajtos at https://github.com/strongloop/loopback-datasource-juggler/pull/1091#issuecomment-248838125.

**Comment by superkhau** [comment]
>IMO, we need a way for connectors to signal what features they support. The shared test suite in juggler should skip tests for features that are not supported.

This is one key point for the DAO architecture discussions that @Amir-61 is working on.

**Comment by superkhau** [comment]
@bajtos @raymondfeng Any final comments before landing?

**Comment by superkhau** [comment]
@duffn Can you help with submitting PRs for the unsupported comment in the downstream connectors (before landing this so at least we have warning messages to cover those areas).

**Comment by bajtos** [comment]
@superkhau I am opposed to landing this patch until we figure out how to keep downstream CI builds of our connectors green. We have invested a lot of time in the recent weeks to get our CI builds green, landing this patch will undo most of that effort.

> I believe you can do similar approach to what @bajtos and I did in [PR#866](https://github.com/strongloop/loopback-datasource-juggler/pull/866).

I am afraid this approach cannot be used easily, because `test/basic-querying.test.js` does not provide any hook for injecting configuration :(

I see the following options how to move forward:

1) Quick hack: attach the information about connector capabilities to the global `getSchema` method. This will make it easy to use `it` or `it.skip` in the tests:

```js
    var itWhenIlikeSupported = getSchema.connectorCapabilities.ilike ? it : it.skip.bind(it);
  
    itWhenIlikeSupported('should support "ilike" that is satisfied', function(done) {
      User.find({where: {name: {ilike: 'john'}}}, function(err, users) {
        if (err) return done(err);
        users.length.should.equal(1);
        users[0].name.should.equal('John Lennon');
        done();
      });
    });
```

Alternatively, require the connectors to define a global object `connectorCapabilities`, similarly how a global `getSchema` is expected.

2) Refactor the shared test suite to export a function that can accept `connectorCapabilities` and `getSchema` as arguments, to avoid global variables. This can be a follow-up story for 1)

3) Alternatively, store the metadata in the connector/datasource instance. Define a new property `dataSource.capabilities` or `dataSource.feature` that can hold information about additional capabilities supported by the connector. We already have [`dataSource.getType()`](https://github.com/strongloop/loopback-datasource-juggler/blob/24f000d9781cc93db5ccdd782bdc9b8507688f9e/lib/datasource.js#L708-L715) that's backed by connector's [prototype.getType()](https://github.com/strongloop/loopback-datasource-juggler/blob/24f000d9781cc93db5ccdd782bdc9b8507688f9e/lib/connectors/memory.js#L51-L53). The downside is that we won't be able to write `it.skip` tests, because the datasource is created only after all tests have been already declared.

I am personally voting for 1).

@superkhau @raymondfeng @Amir-61 thoughts?

**Comment by superkhau** [comment]
>I am opposed to landing this patch until we figure out how to keep downstream CI builds of our connectors green. We have invested a lot of time in the recent weeks to get our CI builds green, landing this patch will undo most of that effort.

I agree we do not land this unless CI stays green. I'm also gonna vote for solution 1 as that looks like the smoothest path to landing.

@duffn Thoughts? Would you like to help submit the other PRs based on this approach?

**Comment by duffn** [comment]
Yes, I will try to find the time to submit the additional PRs.

**Comment by duffn** [comment]
Is this acceptable?
https://gist.github.com/duffn/0d9880b04c9cd617a8a36a0fb434ad88

**Comment by bajtos** [comment]
> Is this acceptable?
https://gist.github.com/duffn/0d9880b04c9cd617a8a36a0fb434ad88

Looks good to me.

**Comment by Amir-61** [comment]
Sorry for joining late on the conversation.

>>I believe you can do similar approach to what @bajtos and I did in PR#866.

>I am afraid this approach cannot be used easily, because test/basic-querying.test.js does not provide any hook for injecting configuration :(

Oh you are absolutely right!

To solve the problem I totally agree with your first approach which is explained https://github.com/strongloop/loopback-datasource-juggler/pull/1091#issuecomment-249199382

**Comment by duffn** [comment]
connectorCapabilities PR submitted to downstream connector.
- [ ] loopback-connector-db2
- [ ] loopback-connector-mysql
- [ ] loopback-connector-cloudant
- [ ] loopback-connector-dashdb
- [ ] loopback-connector-db2z
- [ ] loopback-connector-informix
- [ ] loopback-connector-kv-redis
- [ ] loopback-connector-mongodb
- [ ] loopback-connector-mssql
- [ ] loopback-connector-oracle
- [ ] loopback-connector-postgresql
- [ ] loopback-connector-remote
- [ ] loopback-connector-rest
- [ ] loopback-connector-sqlite3
- [ ] loopback-ibmdb

connectorCapabilities PR accepted to downstream connector.
- [ ] loopback-connector-db2
- [ ] loopback-connector-mysql
- [ ] loopback-connector-cloudant
- [ ] loopback-connector-dashdb
- [ ] loopback-connector-db2z
- [ ] loopback-connector-informix
- [ ] loopback-connector-kv-redis
- [ ] loopback-connector-mongodb
- [ ] loopback-connector-mssql
- [ ] loopback-connector-oracle
- [ ] loopback-connector-postgresql
- [ ] loopback-connector-remote
- [ ] loopback-connector-rest
- [ ] loopback-connector-sqlite3
- [ ] loopback-ibmdb

**Comment by bajtos** [comment]
@slnode ok to test

**Comment by bajtos** [comment]
As far as I am concerned, the code change in this pull request looks good enough to me. @superkhau could you please work with @duffn to tie up any remaining loose ends? Should we mention the two new test flags somewhere in the documentation for building connectors?

**Comment by superkhau** [comment]
>Should we mention the two new test flags somewhere in the documentation for building connectors?

Yes we should. :+1: 

cc @crandmck

**Comment by superkhau** [comment]
@duffn Can you rebase and squash before we land this? Let's try to land those other related PRs first though.

**Comment by duffn** [comment]
Alright, I think I got all of the PRs submitted.

**Comment by superkhau** [comment]
@duffn I managed to land and backport all the connector capability PRs. Can you rebase this PR and lets see how CI fares.

**Comment by duffn** [comment]
Fingers crossed.

**Comment by superkhau** [comment]
@duffn LGTM, gonna land this as the failures are known failures for those downstream connectors. Thank you for the huge contribution. ;)

**Comment by superkhau** [comment]
@duffn Can you submit another PR to loopback to document the two new flags? See https://github.com/strongloop/loopback.io/

@crandmck Can you guide @duffn to where he can add those notes?

**Comment by crandmck** [comment]
@duffn 

Please do update the docs; you can open a PR for docs just as for the code.

Docs for ilike and nilike operators: http://loopback.io/doc/en/lb2/Where-filter.html#operators - source in https://github.com/strongloop/loopback.io/blob/gh-pages/pages/en/lb2/Where-filter.md 

Please at least add it to the table in that section.  If there's more to say about usage, you can also add a subsection as there is for AND/OR and regexp.

Ideally, you would also add an example in http://loopback.io/doc/en/lb2/Where-filter.html#examples

> Should we mention the two new test flags somewhere in the documentation for building connectors?

If you want to address this as well, we'd probably add a new section to http://loopback.io/doc/en/lb2/Building-a-connector.html - source in https://github.com/strongloop/loopback.io/blob/gh-pages/pages/en/lb2/Building-a-connector.md.

Don't worry to much about wording and formatting.  I can always clean it up afterward.

Thanks!!

**Pull request merged**